File: app14_LUA_UNITY/backend/db/migrations/006_create_cross_references.lua
local schema = require("lapis.db.schema")
local types = schema.types
local db = require("lapis.db")

return {
  up = function()
    schema.create_table("cross_references", {
      { "id", types.serial({ primary_key = true }) },
      { "source_threat_code", types.varchar({ null = false }) },
      { "target_threat_code", types.varchar({ null = false }) },
      { "target_threat_title", types.varchar({ null = false }) },
      { "relationship_type", types.varchar({ null = false }) },
      { "description", types.text({ null = false, default = "" }) }
    })

    db.query("alter table cross_references add constraint cross_references_target_fk " ..
      "foreign key (target_threat_code) references threats (code)")
    db.query("create index cross_references_source_idx on cross_references (source_threat_code)")
  end,
  down = function()
    schema.drop_table("cross_references")
  end
}

File: app13_ruby_FastApi/backend/db/migrations/006_create_cross_references.rb
# frozen_string_literal: true

Sequel.migration do
  change do
    create_table(:cross_references) do
      primary_key :id
      foreign_key :source_threat_code, :threats, key: :code, type: String, null: false
      String :target_threat_code, null: false
      String :target_threat_title, null: false
      String :relationship_type, null: false
      String :description, text: true, null: false

      constraint(:cross_references_relationship_type_check) do
        Sequel.lit("relationship_type IN ('equivalent', 'related', 'parent_child', 'maps_to')")
      end

      index :source_threat_code
      index %i[source_threat_code target_threat_code], unique: true, name: :cross_references_unique_pair
    end
  end
end
